store/src/java/com/zimbra/cs/html/OwaspDefang.java

```diff
@@ -0,0 +1,475 @@
+ * See the GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <https://www.gnu.org/licenses/>.
+ * ***** END LICENSE BLOCK *****
+ */
+package com.zimbra.cs.html;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.Writer;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang.StringEscapeUtils;
+import org.owasp.html.AttributePolicy;
+import org.owasp.html.CssSchema;
+import org.owasp.html.ElementPolicy;
+import org.owasp.html.Handler;
+import org.owasp.html.HtmlPolicyBuilder;
+import org.owasp.html.HtmlStreamRenderer;
+import org.owasp.html.PolicyFactory;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.CharStreams;
+import com.zimbra.common.localconfig.DebugConfig;
+import com.zimbra.common.util.StringUtil;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.servlet.ZThreadLocal;
+
+public class OwaspDefang extends AbstractDefang {
+
+    List<String> result = new ArrayList<String>();
+    ExecutorService executor = Executors.newFixedThreadPool(10);
+
+    @Override
+    public void defang(InputStream is, boolean neuterImages, Writer out) throws IOException {
+        String html = CharStreams.toString(new InputStreamReader(is, Charsets.UTF_8));
+        String sanitizedHtml = runSanitizer(html, neuterImages);
+        out.write(sanitizedHtml);
+        out.close();
+    }
+
+    @Override
+    public void defang(Reader reader, boolean neuterImages, Writer out) throws IOException {
+        String html = CharStreams.toString(reader);
+        String sanitizedHtml = runSanitizer(html, neuterImages);
+        out.write(sanitizedHtml);
+        out.close();
+    }
+
+    private String runSanitizer(String html, boolean neuterImages) {
+        Callable<String> task = new OwaspHtmlSanitizer(html, neuterImages);
+        Future<String> future = executor.submit(task);
+        String sanitizedHtml = null;
+        try {
+            // give 10 seconds to sanitize html
+            // TODO: revise the time given
+            sanitizedHtml = future.get(10, TimeUnit.SECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            ZimbraLog.soap.error("Exception during HTML sanitization", e);
+            return null;
+        }
+        executor.shutdown();
```

Comment:
If the service is shutdown after just single task execution, all the other tasks will be rejected. Shutdown should be called when application is getting shutdown.